add nlr check to reorganize function

orpheus opened the pull request.

orpheus committed:
add nlr check to reorganize function

orpheus commented on lib/blockchain/chain.js:
```diff
@@ -855,6 +855,37 @@ class Chain extends AsyncEmitter {
+          tip.height - fork.height
+        );
+
+        const entryWithMostWork = (tip.height > competitor.height) ? tip : competitor
```
isn't this always going to be `competitor`?

bitspill replied:
No, the work value of a block depends upon the difficulty not only the height

`const entryWithMostWork = tip.chainwork.gt(competitor.chainwork) ? tip : competitor`

orpheus replied:
okay I ended up switching this to just using the `competitor` instead of `entryWithMostWork`. Assuming that the walk down always happens from the longer chain which the competitor should always be if a reorg was triggered.

Here too, if the walk down is done on `pindexMostWork` and pindexMostWork is the tip of the incoming fork being considered, `competitor` is the equivalent.

https://github.com/floblockchain/flo/commit/2fe045194caf8137878128ebf01461b9a99e7fb4#diff-24efdb00bfbe56b140fb006b562cc70bR2328

orpheus committed:
refactor/update nlr function; check nlr for spv wallet

bitspill commented on lib/blockchain/chain.js:
```diff
@@ -950,6 +956,46 @@ class Chain extends AsyncEmitter {
+        // mark invalid_child from tip of fork to second block of fork
+        while (competitor.height > fork.height + 2) {
+          const previous = await this.getPrevious(competitor)
+          this.invalidate(previous.hash)
```
I think `invalidate` needs an await since it's also an async function pending a lock on the chain

bitspill commented on lib/blockchain/chain.js:
```diff
@@ -950,6 +956,46 @@ class Chain extends AsyncEmitter {
+
+        // mark invalid first block of fork
+        const previous = await this.getPrevious(competitor)
+        this.invalidate(previous.hash)
```
Why is this invalidation not a part of the above loop?

orpheus replied:
I was following the pattern implemented in flocore. I don't know why it's done this way

https://github.com/floblockchain/flo/commit/2fe045194caf8137878128ebf01461b9a99e7fb4#diff-24efdb00bfbe56b140fb006b562cc70bR2338

bitspill replied:
Ah, in flocore they're being marked different types of invalid.

The first block of the fork is marked as a "failed block" and all subsequent blocks are marked "failed child" it doesn't look like fcoin distinguishes between the two so the separation seemed odd at first impressions

orpheus replied:
makes sense. I'll change the while loop to just run down the whole thing. 
and am I wrong in thinking this?

```
while (pindexWalk->nHeight != pindexFork->nHeight+2) {
            pindexWalk = pindexWalk->pprev;
```

the first two lines of the invalidation while loop start with taking the `pprev` of the `pindexWalk` and invalidating from there. what about `pindexWalk` itself. Doesn't that have to get invalidated (the tip)? Doesn't it seem like were skipping that by starting with `pindexWalk -> prev`

bitspill commented on lib/blockchain/chain.js:
```diff
@@ -855,6 +855,9 @@ class Chain extends AsyncEmitter {
     assert(fork, 'No free space or data corruption.');
 
+    // Check NLR
+    if (this.noLongReorg(tip, fork, competitor)) return true
```
Should there be an await on the `noLongReorg` call?

bitspill commented on lib/blockchain/chain.js:
```diff
@@ -912,6 +915,9 @@ class Chain extends AsyncEmitter {
     assert(fork, 'No free space or data corruption.');
 
+    // Check NLR
+    if (this.noLongReorg(tip, fork, competitor)) return true
```
Same question about await here again

orpheus committed:
add missing awaits

orpheus committed:
fix/update logger warning and fix incorrect assert usage

orpheus committed:
update invalidation stepping to use just the while loop and add missing await to assert check

OstlerDev reviewed (approved):
These changes look good to me. As far as I can tell, it seems to match the functionality that exists currently in FLOCore.

You have my Approval!

orpheus committed:
pull loop variable outside for reassignment

orpheus committed:
setInvalid instead of invalidate + strictEqual assert

instead of running this.invalidate which deadlocks the chain, set the hash to invalid manually with this.setInvalid.

update assertion to use strict equal

orpheus committed:
set competitor chain tip to invaild

setting the competitor chain tip to invaild fixes the problem where the chain continues to sync after nlr. main chain will no longer sync with competitor after nlr

bitspill reviewed (approved):
Don't notice anything stand out wrong, if you've tested a few reorg cases looks good to me.

cchrysostom reviewed (approved):
Reviewed
